### ISSUE TITLE
Rule proposal: const-only-if-immutable

### ISSUE THREAD

**Issue by chomosuke** [typescript-eslint/typescript-eslint#4727]
### Description

Some people (including me) think that the `const` keyword in javascript are misleading, this is because even if an variable is declared as `const`, it can still be modified.

What I would like to do is to enforce `const` if and only if a variable is truly immutable (i.e. having a type that's number, string, boolean or some sort of other immutable types). Otherwise enforce `let` to signify that the variable is mutable.

I'm happy to write a PR to implement this.

### Fail

```ts
// immutable type that are never modified should use const
let a: string | undefined = 'me';

// mutable type should always use let
const b: { a: string } = {
    a: 'me',
};
const c: [string] = ['me'];
```

### Pass

```ts
// immutable type that are never modified should use const
const a: string | undefined = 'me';

// mutable type should always use let
let b: { a: string } = {
    a: 'me',
};
let c: [string] = ['me'];

// immutable type that are modified will be ignored
let d: string = 'me';
d = 'hi';
```

**Comment by bradzacher** [comment]
Thanks for the issue!
This is a very personal, stylistic choice, as such I don't think that such a rule belongs within our general purpose plugin.

Feel free to build on top of our tooling! Perhaps a plugin like `eslint-plugin-functional` might be a good place? Failing that you can always create a rule locally or publish your own plugin!
